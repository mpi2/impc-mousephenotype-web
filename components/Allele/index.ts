export { default as AlleleMap } from './AlleleMap';
export { default as Crispr } from './Crispr';
export { default as ESCell } from './ESCell';
export { default as IVP } from './IVP';
export { default as Mice } from './Mice';
export { default as QCModal } from './QCModal';
export { default as TVP } from './TVP';
