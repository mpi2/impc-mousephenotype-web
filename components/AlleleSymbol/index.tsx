import { formatAlleleSymbol } from "@/utils";

const AlleleSymbol = ({ symbol, withLabel = true }) => {
  const allele = formatAlleleSymbol(symbol);
  return <span>{!!withLabel ? 'Allele: ' : ''}{allele[0]}<sup>{allele[1]}</sup></span>
}

export default AlleleSymbol;