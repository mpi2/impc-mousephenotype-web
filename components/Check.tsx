import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck, faXmark } from "@fortawesome/free-solid-svg-icons";

const Check = ({ isChecked }) => {
  const styles = isChecked ? {} : { backgroundColor: '#8E8E8E' }
  return (
    <span style={styles}>
      <FontAwesomeIcon icon={isChecked ? faCheck : faXmark} style={{ color: '#000' }} />
    </span>
  )
};

export default Check;
