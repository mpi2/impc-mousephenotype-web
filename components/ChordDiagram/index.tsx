import * as d3 from 'd3';
import { useEffect, useRef } from "react";
import styles from './styles.module.scss';

interface Props  {
  // data: Array<Array<number>>;
  width: number;
  height: number;
  topTerms?: Array<string>;
}

function groupTicks(d, step) {
  var k = (d.endAngle - d.startAngle) / d.value;
  return d3.range(0, d.value, step).map(function (value) {
    return {value: value, angle: value * k + d.startAngle, index: d.index};
  });
}

const ChordDiagram = ({ width, height, topTerms }: Props) => {
  const ref = useRef();
  const outerRadius = Math.min(width, height) * 0.5 - 200;
  const innerRadius = outerRadius - 30;
  // TODO: to be removed after getting payload for landing pages
  const matrix = [
    [
      590,
      268,
      712,
      64,
      45
    ],
    [
      268,
      0,
      26,
      12,
      4
    ],
    [
      712,
      26,
      0,
      11,
      6
    ],
    [
      64,
      12,
      11,
      0,
      0
    ],
    [
      45,
      4,
      6,
      0,
      0
    ]
  ]
  let labels = "[{\"name\":\"cardiovascular system phenotype\",\"geneCount\":1622}, {\"name\":\"vision\\\/eye phenotype\",\"geneCount\":268}, {\"name\":\"growth\\\/size\\\/body region phenotype\",\"geneCount\":712}, {\"name\":\"embryo phenotype\",\"geneCount\":64}, {\"name\":\"muscle phenotype\",\"geneCount\":45}]";
  let labels_json = JSON.parse(labels);

  useEffect(() => {
    const svgElement = d3.select(ref.current);

    function fade(opacity) {
      return function (d, i) {
        const { index } = i;
        // hide other chords on mose over
        svgElement.selectAll("path.chord")
          .filter(chord => chord.source.index !== index && chord.target.index !== index)
          .transition()
          .style("stroke-opacity", opacity)
          .style("fill-opacity", opacity);
      };
    }
    const chord = d3.chord()
      .padAngle(0.05)
      .sortSubgroups(d3.descending);

    const arc = d3.arc().innerRadius(innerRadius).outerRadius(outerRadius);

    const ribbon = d3.ribbon().radius(innerRadius);

    const color = d3.scaleOrdinal()
      .domain(d3.range(4))
      .range([
        "rgb(239, 123, 11)",
        "rgb(9, 120, 161)",
        "rgb(119, 119, 119)",
        "rgb(238, 238, 180)",
        "rgb(36, 139, 75)",
        "rgb(191, 75, 50)",
        "rgb(255, 201, 67)",
        "rgb(191, 151, 50)",
        "rgb(239, 123, 11)",
        "rgb(247, 157, 70)",
        "rgb(247, 181, 117)",
        "rgb(191, 75, 50)",
        "rgb(151, 51, 51)",
        "rgb(144, 195, 212)"
      ]);

    const g = svgElement.append("g")
      .attr("transform", `translate(${width / 2}, ${height / 2})`)
      .datum(chord(matrix));

    const group = g.append("g")
      .attr("class", "groups")
      .selectAll("g")
      .data(chords => chords.groups)
      .enter().append("g")
      .attr("class", "group")
      .on("mouseover", fade(.02))
      .on("mouseout", fade(.80));

    group.append("path")
      .style("fill", d => color(d.index))
      .style("stroke", d => d3.rgb(color(d.index)).darker())
      .attr("d", arc);

    const groupTick = group.selectAll(".group-tick")
      .data(d => groupTicks(d, 1e3))
      .enter().append("g")
      .attr("class", "group-tick")
      .attr("transform", d => `rotate(${d.angle * 180 / Math.PI - 90}) translate(${outerRadius},0)`);

    groupTick.append("line")
      .attr("x2", 6);

    groupTick
      .filter(d => d.value % 5e3 === 0)
      .append("text")
      .attr("x", 8)
      .attr("dy", ".35em")
      .attr("transform", d => d.angle > Math.PI ? "rotate(180) translate(-16)" : null)
      .style("text-anchor", d => d.angle > Math.PI ? "end" : null)
      .text(d => labels_json[d.index].name.replace("phenotype", ""));

    g.append("g")
      .attr("class", "ribbons")
      .selectAll("path")
      .data(chords => chords)
      .enter()
      .append("path")
      .attr("d", ribbon)
      .attr("class", "chord")
      .style("fill", d => color(d.target.index))
      .style("stroke", d =>d3.rgb(color(d.target.index)).darker())
      .style("visibility", d => {
        if (topTerms && topTerms.length > 0) {
          if (topTerms.indexOf(labels_json[d.source.index].name) < 0 && topTerms.indexOf(labels_json[d.target.index].name) < 0) {
            return "hidden";
          } else {
            return "visible";
          }
        } else {
          return "visible";
        }
      })
      .append("title").text(d => {
        return d.source.value + " genes present " + labels_json[d.source.index].name + " and " + labels_json[d.target.index].name + ", " ;
      });

  }, []);

  return (
    <div className={styles.wrapper}>
      <svg
        width={width}
        height={height}
        ref={ref}
      />
    </div>
  )
};

export default ChordDiagram;