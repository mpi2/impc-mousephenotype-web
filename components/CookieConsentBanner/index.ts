import dynamic from "next/dynamic";

export default dynamic(() => import("./CookieConsentBanner"), {
  ssr: false,
});
