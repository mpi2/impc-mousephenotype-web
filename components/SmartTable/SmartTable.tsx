import { Form } from "react-bootstrap";
import SortableTable from "@/components/SortableTable";
import Pagination from "@/components/Pagination";
import React, { ReactElement, useEffect, useState } from "react";
import { TableCellProps } from "@/models/TableCell";
import _ from "lodash";
import type { Model } from "@/models";
import Skeleton from "react-loading-skeleton";


const SmartTable = <T extends Model>(props: {
  columns: Array<{
    width: number;
    label: string;
    field?: keyof T,
    cmp: ReactElement<TableCellProps<T>>,
    // refers to disable sort functionality
    disabled?: boolean
  }>,
  data: Array<T>,
  defaultSort: [string, "asc" | "desc"],
  zeroResulsText?: string;
  filterFn?: (item: T, query: string) => boolean,
  additionalTopControls?: ReactElement,
  additionalBottomControls?: ReactElement,
  filteringEnabled?: boolean,
  // set this to false if you need more specific filtering, check All Phenotypes section
  customFiltering?: boolean,
  showLoadingIndicator?: boolean,
  customSortFunction?: (data: Array<T>, field: string, order: "asc" | "desc") => Array<T>;
  highlightRowFunction?: (item: T) => boolean;
  highlightRowColor?: string;
  pagination?: {
    totalItems: number;
    onPageChange: (newPage: number) => void;
    onPageSizeChange: (newPageSize: number) => void;
    page: number;
    pageSize: number;
  };
  onSortChange?: (sortOptions: string) => void;
}) => {
  const {
    filteringEnabled = true,
    customFiltering = false,
    zeroResulsText = 'No data available',
    showLoadingIndicator = false,
    highlightRowFunction = () => false,
    highlightRowColor = '#00b0b0',
    pagination = null,
    onSortChange = (_) => {},
  } = props;
  const [query, setQuery] = useState(undefined);
  const [sortOptions, setSortOptions] = useState<string>('');

  const internalShowFilteringEnabled = filteringEnabled && !!props.filterFn && !customFiltering;

  let mutatedData = props.data || [];
  if (props.filterFn) {
    mutatedData = mutatedData?.filter(item => props.filterFn(item, query));
  }
  const [field, order] = sortOptions.split(';');
  if (field && order && !!pagination) {
    mutatedData = !!props.customSortFunction
      ? props.customSortFunction(mutatedData, field, order as "asc" | "desc")
      : _.orderBy(mutatedData, field, order as "asc" | "desc")
  }

  useEffect(() => {
    if (sortOptions) {
      onSortChange(sortOptions);
    }
  }, [sortOptions]);

  const additionalControls = internalShowFilteringEnabled ?  (
    <Form.Control
      type="text"
      style={{
        display: "inline-block",
        width: 200,
        marginRight: "2rem",
      }}
      aria-label="Filter by parameters"
      id="parameterFilter"
      className="bg-white"
      placeholder="Search "
      onChange={(el) => {
        setQuery(el.target.value.toLowerCase() || undefined);
      }}
    ></Form.Control>
  ) : props.additionalTopControls

  return (
    <Pagination
      data={mutatedData}
      additionalTopControls={additionalControls}
      additionalBottomControls={props.additionalBottomControls}
      controlled={!!pagination}
      {...pagination}
    >
      {(pageData) => (
        <SortableTable
          doSort={([field, order]) => {
            setSortOptions(`${field};${order}`);
          }}
          defaultSort={props.defaultSort}
          headers={props.columns.map(
            ({ field, cmp, ...rest }) => ({
              ...rest,
              field: field as string,
            }))
          }
        >
          {pageData.map((d, index) => (
            <tr key={index} style={highlightRowFunction(d) ? { border: `3px solid ${highlightRowColor}` } : {}}>
              {props.columns.map(({ field, cmp }, index) => (
                <td key={index} style={{ borderColor: `var(--bs-table-border-color)` }}>
                  {React.cloneElement(cmp, { value: d, field })}
                </td>
              ))}
            </tr>
          ))}
          {(pageData.length === 0 && showLoadingIndicator) && (
            <tr>
              {props.columns.map((_, index) => (
                <td key={index}>
                  <Skeleton />
                </td>
              ))}
            </tr>
          )}
          {(pageData.length === 0 && !showLoadingIndicator) && (
            <tr>
              <td colSpan={props.columns.length}>
                <b>{zeroResulsText}</b>
              </td>
            </tr>
          )}
        </SortableTable>
      )}
    </Pagination>
  )
};

export default SmartTable;