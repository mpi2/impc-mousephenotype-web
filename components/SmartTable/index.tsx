
export { default as SmartTable } from './SmartTable';
export { default as SimpleTextCell } from './Cells/SimpleText';
export { default as LinkCell } from './Cells/Link';
export { default as TransformCell } from './Cells/Transform';