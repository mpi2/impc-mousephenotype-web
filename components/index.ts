export { default as Card } from './Card';
export { default as Search } from './Search';
export { default as AlleleSymbol } from './AlleleSymbol';
export { default as DownloadData } from './DownloadData'