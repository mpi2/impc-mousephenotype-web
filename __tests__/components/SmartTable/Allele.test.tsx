import { render } from '@testing-library/react';
import { AlleleCell } from "@/components/SmartTable";

describe('Allele cell component', () => {
  it('renders correctly', () => {
    const data = { allele: 'Nxn<em1(IMPC)Mbp>' };
    const { container } = render(<AlleleCell value={data} field="allele" />);
    expect(container).toMatchSnapshot();
  });
});