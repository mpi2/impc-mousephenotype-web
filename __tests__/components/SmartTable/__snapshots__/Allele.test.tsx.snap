// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allele cell component renders correctly 1`] = `
<span>
  <i>
    Nxn
    <sup>
      em1(IMPC)Mbp
    </sup>
  </i>
</span>
`;
